### PR TITLE
feat(handler): regex commands

### PR DESCRIPTION
**Pull request by almostSouji**
This PR aims to introduce the ability for commands to be executed via regular expressions.

- Command: new optional regExp property which can hold a regular expression, the command can be triggered with. The positional args are retrieved from the capture groups
- Command#execute has a new argument with the execution context (enum, so far prefixed and regexp, other - like web - are thinkable) allowing for context-aware execution of commands (silent fails on regex commands to not disturb conversations if args are only later found to be unusable for this command, due to accidental matching.
- Command: consumes the Args instance as it would with other commands, flags and options are omitted, UX seems to suffer once one introduces flags into pattern commands that can execute on any position in the message body
- Command parsing: slight restructure to allow attempting to execute via regex should no prefix or command name match be found
- After the first matching pattern no further patterns will be tested

<details>
<summary>first idea (revisited after discussion below)</summary>

To facilitate this re-arranging command parsing and adding to the Command interface was required:

- **Command** | Command has a new (optional) regExp property which can hold a regular expression the command can be triggered with
- **Command** | Command#execute has a new (optional) argument with the capability to provide additional information in special circumstances (an interface rather than a single value was chosen to make this concept easily extendable should need arise). This is used with regular expression commands to provide the match, preventing the need to re-do the matching and giving information about the special circumstance (will be a non-null RegExpExecArray if executes through a regular expression)
- **Parsing** | Allow further checks if no matching command is found
- **Parsing** | Re-parse the message content without a prefix to be able to still provide a valid Args instance and not drop flags, should they be required for the command.
- **Parsing** | If any command is found no regular expressions will be checked
- **Parsing** | After the first matching expression no further expressions are checked

</details>

### CODE REVIEW

**Comment by appellation** [comment]
While I'm in favor of this concept, I'm not immediately onboard with the usage of regular expressions here. Passing the matches to the command seems like a hack that offloads parsing problems onto the command implementation itself. I would much rather see commands export some kind of custom lexer/parser to handle their specific needs. This way commands can consume `Args` without being aware of the parsing involved in their derivation.

File: src/handler/bin/index.ts
Comment (NotSugden) [inline src/handler/bin/index.ts:114]: is there a missing else here or..? never seen this before

Comment (almostSouji) [reply src/handler/bin/index.ts:114]: Opens a new scope, avoid having to think of names for re-parsing. I don't agree in that avoiding it would "make more sense" per se. However the current approach is unlikely to be introduced anyways, hence the conversion back to draft,

**Comment by almostSouji** [comment]
> While I'm in favor of this concept, I'm not immediately onboard with the usage of regular expressions here. Passing the matches to the command seems like a hack that offloads parsing problems onto the command implementation itself. 

The main idea here is to make the commands context aware of how it was executed, to be able to handle commands differently based on its execution. If executed by a regex the command feedback on failure could be much less verbose (maybe even silent) to prevent disturbing the channel on accidental activation if the pattern is deemed unusable further down in execution where on regular execution the user can be informed as to what exactly caused the failure.

One example would be a `username#tag` combination accidentally triggering the `repo#issue` pattern (but the "repo" being the username would not be in the allowed repositories on this guild so the command could fail silently, while on explicit command execution this can result in something like "Make sure the repository "repo" is enabled on the guild / server".

> I would much rather see commands export some kind of custom lexer/parser to handle their specific needs. This way commands can consume `Args` without being aware of the parsing involved in their derivation.

Would you mind elaborating on this? I'm not sure I understand the approach you propose here. Are you suggesting that each command should have its own parser depending on the "pattern" it requires, with "normal" commands being the intuitive idea of space delimited arguments? I'm not too sure of the benefits here. 

I'm trying to explain my thought process through an example:

Command arguments could vary between usages

- `?issue-pr <repo> <issue>`
- `<repo>#<issue>`

While the first would get the arguments from consuming them from the `Args` instance the latter could retrieve them from the pattern-parsing. This requires minimal effort in the command itself, as just the assignment of the used args varies. If a command is executed normally 2x `.single()` is consumed, if that's not the case we can re-use the parsing pattern from a regular expression, which was used to determine if the command should execute anyways.

The re-parsing without requiring a prefix is primarily done to still have access to other control structures like (option) flags, which still follows the normal command flow (no extra code to handle the case here). Optimally we'd use one parsing flow for this, however if the prefix does not match the flags aren't even parsed, accordingly the tokens not available for constructing an `Args` instance (flags and options probably being the main use case here) thus requiring a re-parsing effort.

**Comment by appellation** [comment]
I can see how being context-aware might be beneficial for the command, but I don't believe that this should come at the cost of putting command arguments in a different place depending on context. Arguments are arguments, regardless of the format from which they were parsed; we should not handle them differently.

The concept I am trying to convey is that, fundamentally, the command itself should not have to understand more than the `Args` that are passed into it. `?issue-pr <repo> <issue>` and `<repo>#<issue>` are formatting details that the command should not know about. *Command execution should not change based on input format.* I can see how having additional context may be useful, but this should be completely separated from the format of the data provided to the command execution.

My original comment was intentionally vague in this area because I do not know how to use lexure to parse `<repo>#<issue>` into an `Args` instance. I am commenting simply because I believe this is the correct approach, not because I know the technical solution.

**Comment by almostSouji** [comment]
thank you, now I understood the main concern, I believe the following approach should address this:

1. try to parse the command normally through lexure, if this results in a valid command, execute it
2. should this not be possible iterate through the commands and attempt to parse the message content against a command-specific regex
3. should this succeed fabricate Args from the parsed pieces
4. re-apply flags to this fabricated Args instance
5. execute the command, pass an indicator to the command so the flow can be handled differently. *

\* I believe it is crucial to user experience to handle the command flow (especially on rejecting) different if the command is executed as a pattern, you'd not want `username#discrim` to prompt a response along the lines of "this is not a valid repository"

as for (4) i'm not yet entirely sure how to achieve this yet, but should be possible one way or another

That way the command will just consume Args (either the "natural" or "fabricated" structure) and won't have to differ in execution apart from the aforementioned silent rejection.

**Comment by appellation** [comment]
That seems like a reasonable approach. I would prefer to avoid regular expressions entirely if possible, but perhaps that's excessively complicated.

I agree that the context is valuable, so perhaps some sort of enum flag can be passed to the command to determine the "source" of the args.

**Comment by almostSouji** [comment]
I have made some changes to hopefully reflect the discussion above and avoid some of the unwanted design patterns and caveats. Maybe the move from RegExp to custom parsers can be made in a later update?

File: src/handler/src/Command.ts
Comment (iCrawl) [inline src/handler/src/Command.ts:17]: Any specific reason for the init with 0?

Comment (almostSouji) [reply src/handler/src/Command.ts:17]: Really just verbosity as per <https://www.typescriptlang.org/docs/handbook/enums.html#numeric-enums> numeric and 0-indexed is the default anyways, if you feel it's cleaner without the init i'll of course gladly remove it or change it to the respective "flag" string, if that's preferred

**Review by appellation — APPROVED** [review]
This looks good to me. I would suggest removing the enum initialization, but wait for a review from @iCrawl.

File: src/handler/src/Command.ts
Comment (iCrawl) [reply src/handler/src/Command.ts:17]: Yeah, we don't need that. If its 0-indexed we can just remove the initialization.

Comment (almostSouji) [reply src/handler/src/Command.ts]: 👍  done in https://github.com/Naval-Base/yuudachi/pull/775/commits/3a90be1a65701a33051bc176fa6ec68d328b6cde

**Review by iCrawl — APPROVED** [review]

**Pull request merged**
